src/MongoDB.Analyzer/Core/Builders/BuildersResolveVariables.cs

```diff
@@ -0,0 +1,445 @@
+            }
+        }
+        else if (rightHandSide is InvocationExpressionSyntax invocationExpression)
+        {
+            canEvaluate = builderToAnalysisContext.TryGetValueSafe(invocationExpression, out var expressionAnalysisContext);
+
+            if (canEvaluate)
+            {
+                syntaxNodeToVariableValueMap ??= new Dictionary<SyntaxNode, VariableValue>();
+                var rightVariableValue = new VariableValue(expressionAnalysisContext, expressionAnalysisContext.Node.Locations.ToList());
+                syntaxNodeToVariableValueMap.Add(invocationExpression, rightVariableValue);
+            }
+        }
+        else if (rightHandSide is BinaryExpressionSyntax binaryExpressionSyntax)
+        {
+            canEvaluate = builderToAnalysisContext.TryGetValueSafe(binaryExpressionSyntax, out var expressionAnalysisContext);
+
+            if (!canEvaluate)
+            {
+                var leftEvaluationResult = PreEvaluate(binaryExpressionSyntax.Left, contextInformation, syntaxNodeToVariableValueMap : null, displayDiagnostics);
+
+                if (leftEvaluationResult.CanEvaluate)
+                {
+                    var rightEvaluationResult = PreEvaluate(binaryExpressionSyntax.Right, contextInformation, syntaxNodeToVariableValueMap : leftEvaluationResult.SyntaxNodeToVariableValueMap, displayDiagnostics);
+                    canEvaluate = leftEvaluationResult.CanEvaluate && rightEvaluationResult.CanEvaluate;
+                    syntaxNodeToVariableValueMap = rightEvaluationResult.SyntaxNodeToVariableValueMap;
+                }
+            }
+            else
+            {
+                syntaxNodeToVariableValueMap ??= new Dictionary<SyntaxNode, VariableValue>();
+                var rightVariableValue = new VariableValue(expressionAnalysisContext, expressionAnalysisContext.Node.Locations.ToList());
+                syntaxNodeToVariableValueMap.Add(binaryExpressionSyntax, rightVariableValue);
+            }
+        }
+
+        if (isCompoundExpression && canEvaluate)
+        {
+            var assignmentExpression = syntaxNode as AssignmentExpressionSyntax;
+            var leftHandSideIdentifier = assignmentExpression.Left.TrimParenthesis() as IdentifierNameSyntax;
+            canEvaluate &= variableValues.TryGetValueSafe(leftHandSideIdentifier.Identifier.ValueText, out var leftVariableValue);
+
+            if (canEvaluate)
+            {
+                syntaxNodeToVariableValueMap ??= new Dictionary<SyntaxNode, VariableValue>();
+                syntaxNodeToVariableValueMap.Add(leftHandSideIdentifier, leftVariableValue);
+            }
+        }
+
+        return new EvaluationResult(canEvaluate, syntaxNodeToVariableValueMap);
+    }
+
+    private static ExpressionAnalysisContext GenerateNewExpressionAnalysisContext(
+        SyntaxNode syntaxNode,
+        Dictionary<SyntaxNode, VariableValue> builderNodeToVariableValue)
+    {
+        var syntaxKind = syntaxNode.Kind();
```

Comment:
no need for dedicated syntaxKind variable. Seems that IsCompoundExpression can just receive syntaxNode

Reply:
Done